# Data 200: Data Systems for Data Analytics (Spring 2025)

# Homework 10: Database Programming

<font color='red'>**Due Date and Time:** 11:59pm on Tuesday, 3/31/2025 </font>
---
Enter your name in the markdown cell below.

# Name: Thao 

In [1]:
## RUN THIS CELL TO GET THE RIGHT FORMATTING AND TO LOAD NumPy
import requests
import numpy as np
from IPython.core.display import HTML
styles = requests.get("https://raw.githubusercontent.com/Harvard-IACS/2018-CS109A/master/content/styles/cs109.css").text
HTML(styles)

# Tasks

- Review pages 249-280 in the Course Notes.
- Complete the **Applying Filtering, Ordering and Grouping to Queries** and **Advanced SQLAlchemy Queries** chapters of the **Introduction to Databases in Python** course on DataCamp.-
- OPTIONAL: **Customizing Seaborn Plots** and **Additional Plot Types** chapters of the **Intermediate Data Visualization with Seaborn** course on DataCamp.
- Submit your completed Jupyter Notebook on Moodle.

# Exercises (Part 1)

This homework will make use of both the `book.db` and `school.db` databases, both of which should be in the `dbfiles` folder after you unzipped this assignment.  **In addition to exercises, this notebook will demonstrate what we learned about database programming!**

**Run the code cell below to import `sqlalchemy`, `pandas`, and `os.path`**

In [2]:
import pandas as pd
import sqlalchemy as sa
import os.path
import sqlite3

**Run the code cell below to create a *connection string* that that can be used to connect to the `book` database using sqlite.**

In [3]:
protocol = "sqlite"
driver = "pysqlite"
dbdir = "dbfiles"
database = "book"

cstring = f"{protocol}+{driver}:///{os.path.join(dbdir, database + '.db')}"
print(cstring)

sqlite+pysqlite:///dbfiles\book.db


**Run the code cell below to establish the connection to the `book` database using the connection string `cstring`.**

In [48]:
engine = sa.create_engine(cstring)
connection = engine.connect()

**Run the code cell below to perform a simple query to obtain all the columns of the `indicates0` table and print the results.**

In [5]:
# Define the query as a Python string
query = "SELECT *  FROM indicators0"

# Execute the query
result_proxy = connection.execute(query)

# Retrieve the entire table resulting from the query
result_list = result_proxy.fetchall()

# Print the results (which is a result proxy)
print(result_list)

[('CHN', 1386.4, 12143.5, 76.4, 1469.88), ('FRA', 66.87, 2586.29, 82.5, 69.02), ('GBR', 66.06, 2637.87, 81.2, 79.1), ('IND', 1338.66, 2652.55, 68.8, 1168.9), ('USA', 325.15, 19485.4, 78.5, 391.6)]


**Run the code cell below to print the number of records returned, and iterate over the list and print the tuple records.**

In [6]:
print('Number of records returned = ', len(result_list), '\n')

for record in result_list:
    print(record)

Number of records returned =  5 

('CHN', 1386.4, 12143.5, 76.4, 1469.88)
('FRA', 66.87, 2586.29, 82.5, 69.02)
('GBR', 66.06, 2637.87, 81.2, 79.1)
('IND', 1338.66, 2652.55, 68.8, 1168.9)
('USA', 325.15, 19485.4, 78.5, 391.6)


**Run the code cell below to perform a multi-line query of the `indicators` table and print the results.  Note the query is contained inside triple-double quotes!**

In [7]:
# Define the query as a Python string
query = """
SELECT code, pop, gdp, life, cell 
FROM indicators
WHERE year = 2017 AND code IN ('CHN', 'IND', 'FRA', 'USA')
"""

# Execute the query
result_proxy = connection.execute(query)

# Retrieve the entire table resulting from the query
result_list = result_proxy.fetchall()

# Print the number of records and iterate over the list to print the records
print('Number of records returned = ', len(result_list), '\n')

for record in result_list:
    print(record)

Number of records returned =  4 

('CHN', 1386.4, 12143.5, 76.4, 1469.88)
('FRA', 66.87, 2586.29, 82.5, 69.02)
('IND', 1338.66, 2652.55, 68.8, 1168.9)
('USA', 325.15, 19485.4, 78.5, 391.6)


**Run the code cell below to close the connection to the database and delete the engine.**

In [8]:
connection.close()

del engine

<div class="exercise"><b>Exercise 1:</b></div> 

Create and **print** a connection string (called `cstring`) that can be used to connect to the `school` database (which should be in the `dbfiles` directory) using sqlite.  This is almost identical to what we just did in relation to the `book` database.

Below is the output from my solution:<br>
<code>
sqlite+pysqlite:///dbfiles\school.db
</code>

In [50]:
protocol = "sqlite"
driver = "pysqlite"
dbdir = "dbfiles"
database = "school"

cstring = f"{protocol}+{driver}:///{os.path.join(dbdir, database + '.db')}"
print(cstring)

sqlite+pysqlite:///dbfiles\school.db


**Run the code cell below to establish the connection using your newly defined connection string `cstring`.**

In [51]:
engine = sa.create_engine(cstring)
connection = engine.connect()

<div class="exercise"><b>Exercise 2:</b></div> 

Write Python code to do the following:<br>

- Define a query string that selects the instructors (last name, first name) teaching the classes with `classid` given by 21014, 21088, and 21256. Specifically, the resulting table should contain the `classid`, `instructorlast`, and `instructorfirst` fields.
- Execute the query and fetch the results.
- Print the number of records and iterate over the list to print the records.

Below is the output from my solution:<br>
<code>
Number of records returned =  3 
</code><code>
(21014, 'Foster', 'Helen')
(21088, 'Hawkins', 'Grace')
(21256, 'Garrett', 'Jason')
</code>

In [11]:
query = """
    SELECT b.classid, a.instructorlast, a.instructorfirst 
    FROM instructors AS a
    INNER JOIN instructor_class AS b 
    ON a.instructorid = b.instructorid
    WHERE classid IN (21014, 21088, 21256)
"""

result_proxy = connection.execute(query)
result_list = result_proxy.fetchall()
print('Number of records returned = ', len(result_list), '\n')

for record in result_list:
    print(record)

Number of records returned =  3 

(21014, 'Foster', 'Helen')
(21088, 'Hawkins', 'Grace')
(21256, 'Garrett', 'Jason')


**Carefully study and run the code cell below where we:**

- Create a query string that requests the subject, course number, section, and meeting time for the classes with IDs 21014, 21088, 21256, and 21444.
- Use the Pandas function `read_sql_query()` to perform the query (over our `connection`) and assign the result to a Pandas data frame called `pandas_table`.
- Print the data frame.

In [12]:
db_path = os.path.join(dbdir, f"{database}.db")

In [13]:
conn = sqlite3.connect(db_path)

In [14]:
query = """
SELECT coursesubject || '-' || coursenum || 
       '-' || classsection
AS class, classmeeting, instructorlast, instructorfirst
FROM classes INNER JOIN instructor_class USING (classid)
INNER JOIN instructors USING (instructorid)
WHERE classid IN (21014, 21088, 21256, 21444)
"""

pandas_table = pd.read_sql_query(query, conn)

print(pandas_table)

         class     classmeeting instructorlast instructorfirst
0  PSYC-100-02  09:30-10:20 MWF         Foster           Helen
1   FYS-102-02   13:30-14:50 TR        Hawkins           Grace
2  ECON-240-01   11:30-13:20 MW        Garrett           Jason
3  PHED-180-01             None          Price          Taylor


<div class="exercise"><b>Exercise 3:</b></div> 

Write Python code to do the following:<br>

- Define a query string that determines the number of instructors with the same first name and order the results in descending order.  Specifically, project the `instructorfirst` column and alias the number of instructors with the same name as `num_name`.  Limit the results to 10 records.
- Execute the query and fetch the results using `pd.read_sql_query()`.
- Print the resulting data frame.

Below is the output from my solution (make sure your output matches):<br>
<code>
  instructorfirst  num_name
0            Kyle         6
1            Adam         6
2        Danielle         4
3          Taylor         3
4       Stephanie         3
5          Sandra         3
6          Robert         3
7         Phillip         3
8            Lori         3
9          Judith         3
</code>

In [15]:
query = """
SELECT instructorfirst, COUNT(instructorfirst) AS num_name
FROM instructors 
GROUP BY instructorfirst
ORDER BY COUNT(instructorfirst) DESC
LIMIT 10
"""
pandas_table = pd.read_sql_query(query, conn)
print(pandas_table)

  instructorfirst  num_name
0            Kyle         6
1            Adam         6
2        Danielle         4
3          Taylor         3
4       Stephanie         3
5          Sandra         3
6          Robert         3
7         Phillip         3
8            Lori         3
9          Judith         3


**Run the code cell below that defines a function, `first_name_list()`, that builds a string query using a variable (`name`) to retrieve a Pandas data frame containing all the instructors (first and last names) with that first name.**

In [16]:
def first_name_list(dbcon, name):
    
    template = """
    SELECT instructorfirst, instructorlast
    FROM instructors
    WHERE instructorfirst = '{}'
    """
    
    query = template.format(name)
    
    df = pd.read_sql_query(query, con=dbcon)

    return df

**Run the code cell below that calls the `first_name_list()` function to find all the instructors with the first name Kyle.**

In [17]:
names = first_name_list(conn, 'Kyle')
print(names)

  instructorfirst instructorlast
0            Kyle          Berry
1            Kyle         Wagner
2            Kyle           Pena
3            Kyle         Rivera
4            Kyle        Russell
5            Kyle        Stevens


<div class="exercise"><b>Exercise 4:</b></div> 

Write a function `departmentsByDivision(dbcon, div)` that queries the `departments` table for the set of departments in the given division. Use Python string composition and the `format()` method for building the query. Return the result as a Pandas data frame.

**Note:** In the second code cell below you will test your function by finding the departments in the Humanities division.  The output from my solution is as follows:<br>
<code>
      departmentname
0  Classical Studies
1            English
2            History
3    Modern Language
4         Philosophy
5           Religion
</code>

In [18]:
def departmentsByDivision(dbcon, div): 
    
    template = """
    SELECT departmentname
    FROM departments 
    WHERE division = '{}'
    """
    query = template.format(div)
    
    df = pd.read_sql_query(query, con=dbcon)

    return df

**Run the code cell below to verify your function works properly.**

In [19]:
dept = departmentsByDivision(conn, 'Humanities')
print(dept)

      departmentname
0  Classical Studies
1            English
2            History
3    Modern Language
4         Philosophy
5           Religion


<div class="exercise"><b>Exercise 5:</b></div> 

Write a function `subjectCourses(dbcon, subject, low, high)` that queries the `courses` table for the courses in the given `subject` whose course number is between `low and high`, inclusive. Use Python string composition and the `format()` method for building the query. Return the result as a Pandas data frame.

**Note:** In the second code cell below you will test your function by finding all the Math courses with course numbers beween 100 and 250.  The output from my solution is as follows:<br>
<code>
                    coursetitle  coursenum
0        Elements of Statistics        102
1        Essentials of Calculus        121
2  Calculus I (Single Variable)        123
3   Calculus II (Multivariable)        124
4        Intro Proof Techniques        210
5              Technical Comm I        215
6                  Calculus III        222
7      Linear Algebra/Equations        231
</code>

In [20]:
def subjectCourses(dbcon, subject, low, high): 
    
    template = """
    SELECT coursetitle, coursenum
    FROM courses
    WHERE coursesubject = '{}' AND coursenum >= {} AND coursenum <= {}
    ORDER BY coursenum
    """
    query = template.format(subject, low, high)
    
    df = pd.read_sql_query(query, con=dbcon)

    return df

**Run the code cell below to verify your function works properly.**

In [21]:
courses_df = subjectCourses(conn, 'MATH', 100, 250)
print(courses_df)

                    coursetitle  coursenum
0        Elements of Statistics        102
1        Essentials of Calculus        121
2  Calculus I (Single Variable)        123
3   Calculus II (Multivariable)        124
4        Intro Proof Techniques        210
5              Technical Comm I        215
6                  Calculus III        222
7      Linear Algebra/Equations        231


# Exercises (Part 2)

**Study and run the code cell below that creates a function `studentsByCity(dbcon, city)` that queries the `students` table to find all the students from the provided city. Note that this makes use of *binding variables*.  Furthermore, note that there is *not* quotes around `:city`, which is one of the advantages of using binding.**

In [29]:
import sqlalchemy as sa
def studentsByCity(dbcon, city):
    
    pyquery = """
    SELECT studentfirst AS first, studentlast AS last, studentcity AS city, studentstate AS state
    FROM students
    WHERE studentcity = '{}'
    """
    
    #prepare_stmt = sa.sql.text(pyquery)
    #bound_stmt = prepare_stmt.bindparams(city_in = city)
    bound_stmt = pyquery.format(city)
    
    print(bound_stmt)
    
    df = pd.read_sql_query(bound_stmt, dbcon)

    return df

**Run the code cell below that calls the `studentsByCity()` function to find all the students who are from Baltimore.**

In [31]:
students_df = studentsByCity(conn, 'Baltimore')
print(students_df)


    SELECT studentfirst AS first, studentlast AS last, studentcity AS city, studentstate AS state
    FROM students
    WHERE studentcity = 'Baltimore'
    
      first       last       city state
0       Léo    Poirier  Baltimore    MD
1   William     Nelson  Baltimore    MD
2  Kimberly       Cook  Baltimore    MD
3      Ryan      Black  Baltimore    MD
4     Ayaan  Bhatnagar  Baltimore    MD
5    Sophie       Vogt  Baltimore    MD
6    Ronald     Vargas  Baltimore    MD
7  Kimberly      McCoy  Baltimore    OH
8  Courtney      Ellis  Baltimore    MD


<div class="exercise"><b>Exercise 6:</b></div> 

Write a function `deparmentsByDivisionBind(dbcon, div)` that once again queries the `departments` table for the set of departments in the given division, but this time use **variable binding to build the query**. Return the result as a Pandas data frame.  *So this is just like Exercise 4, but this time you are using variable binding.*

**Note:** In the second code cell below you will test your function by finding the departments in the Fine Arts division.  The output from my solution is as follows:<br>
<code>
                   departmentname
0  Art History and Visual Culture
1                          Cinema
2                           Dance
3                           Music
4                         Theatre
</code>

In [40]:
def departmentsByDivisionBind(dbcon, div): 
    pyquery = """
    SELECT departmentname
    FROM departments 
    WHERE division = '{}'
    """
    bound_stmt = pyquery.format(div)
    print(bound_stmt)
    df = pd.read_sql_query(bound_stmt, dbcon)
    return df

**Run the code cell below to verify your function works properly.**

In [41]:
depts = departmentsByDivisionBind(conn, 'Fine Arts')
print(depts)


    SELECT departmentname
    FROM departments 
    WHERE division = 'Fine Arts'
    
                   departmentname
0  Art History and Visual Culture
1                          Cinema
2                           Dance
3                           Music
4                         Theatre


<div class="exercise"><b>Exercise 7:</b></div> 

Write a function `subjectCoursesBind(dbcon, subject, low, high)` that once again queries the `courses` table for the courses in the given `subject` whose course number is between `low and high` (inclusively), but this time use **variable binding to build the query**. Return the result as a Pandas data frame.  *So this is just like Exercise 5, but this time you are using variable binding.*

**Note:** In the second code cell below you will test your function by finding all the Econ courses with course numbers beween 200 and 350.  The output from my solution is as follows:<br>
<code>
               coursetitle  coursenum
0         Economic Justice        201
1  Econ Growth/Env Sustain        202
2  Economies in Transition        240
3   Intermediate Macroecon        301
4   Intermediate Microecon        302
5       Intro Econometrics        307
</code>

In [43]:
def subjectCoursesBind(dbcon, subject, low, high): 
    
    pyquery = """
    SELECT coursetitle, coursenum
    FROM courses
    WHERE coursesubject = '{}' AND coursenum >= {} AND coursenum <= {}
    ORDER BY coursenum
    """
    bound_stmt = pyquery.format(subject, low, high)
    print(bound_stmt)
    df = pd.read_sql_query(bound_stmt, dbcon)
    return df

**Run the code cell below to verify your function works properly.**

In [45]:
courses_df = subjectCoursesBind(conn, 'ECON', 200, 350)
print(courses_df)


    SELECT coursetitle, coursenum
    FROM courses
    WHERE coursesubject = 'ECON' AND coursenum >= 200 AND coursenum <= 350
    ORDER BY coursenum
    
               coursetitle  coursenum
0         Economic Justice        201
1  Econ Growth/Env Sustain        202
2  Economies in Transition        240
3   Intermediate Macroecon        301
4   Intermediate Microecon        302
5       Intro Econometrics        307


<div class="exercise"><b>Exercise 8:</b></div> 

In the code cell below, write Python code to close the database connection and delete the engine.

In [49]:
connection.close()
del engine  